### PR TITLE
gh-86298: Ensure that __loader__ and __spec__.loader agree in warnings.warn_explicit()

### PR DESCRIPTION
**Pull request by warsaw**
In `_warnings.c`, in the C equivalent of `warnings.warn_explicit()`, if the module globals are given (and not None), the warning will attempt to get the source line for the issued warning.  To do this, it needs the module's loader.

Previously, it would only look up `__loader__` in the module globals.  In https://github.com/python/cpython/issues/86298 we want to defer to the `__spec__.loader` if available.

The first step on this journey is to check that `loader == __spec__.loader` and issue another warning if it is not.  This commit does that.

Since this is a PoC, only manual testing for now.

```python
# /tmp/foo.py
import warnings

import bar

warnings.warn_explicit(
    'warning!',
    RuntimeWarning,
    'bar.py', 2,
    module='bar knee',
    module_globals=bar.__dict__,
    )
```

```python
# /tmp/bar.py
import sys
import os
import pathlib

# __loader__ = pathlib.Path()
```

Then running this: `./python.exe -Wdefault /tmp/foo.py`

Produces:

```
bar.py:2: RuntimeWarning: warning!
  import os
```

Uncomment the `__loader__ = ` line in `bar.py` and try it again:

```
sys:1: ImportWarning: Module bar; __loader__ != __spec__.loader (<_frozen_importlib_external.SourceFileLoader object at 0x109f7dfa0> != PosixPath('.'))
bar.py:2: RuntimeWarning: warning!
  import os
```

* Issue: gh-86298

Automerge-Triggered-By: GH:warsaw

### CODE REVIEW

**Comment by warsaw** [comment]
Tagging @brettcannon

**Comment by warsaw** [comment]
Converting from draft since I now have tests.

**Comment by warsaw** [comment]
It's better to split these consistency fixes up into separate PRs, so this one only addresses the `_warnings.c` part of the issue.

**Comment by warsaw** [comment]
@brettcannon I think this branch is ready for review.  It only changes the `_warnings.c` case, adds tests, and updates some documentation.  We'll split any related changes outside of this file into separate PRs.

File: Python/_warnings.c [hunk 1]
```diff
@@ -977,25 +977,59 @@ warnings_warn_impl(PyObject *module, PyObject *message, PyObject *category,
...
+    Py_INCREF(module_name);
+
+    /* Check/get the requisite pieces needed for the loader.  Get both the
+       __spec__.loader and __loader__ attributes and warn if they are not the
+       same.
+    */
+    spec = _PyDict_GetItemWithError(module_globals, &_Py_ID(__spec__));
+    if (spec == NULL) {
+        Py_DECREF(module_name);
+        return NULL;
```
Comment (brettcannon) [inline Python/_warnings.c]: Do we want to fail at this point, or do we want to warn that we are falling through to `__loader__`?

**Review by brettcannon — APPROVED** [review]

File: Python/_warnings.c (hunk 1 above)
Comment (warsaw) [reply Python/_warnings.c]: That's a good point.  After discussion and thinking this through, here are the cases I think we should handle:

```
    /* This is part of the migration away from module attributes and into a
       fully __spec__ future.  See GH#97850

       * If you have a __loader__ and a __spec__.loader but they are not the
         same, in Python 3.12 we issue a DeprecationWarning and fall back to
         __loader__ for backward compatibility.  In Python 3.14, we'll flip
         this case to ignoring __loader__ entirely, without error.

       * If you do not have a __spec__ or __spec__.loader, we also issue a
         DeprecationWarning and fall back to __loader__.  In Python 3.14,
         we'll make this case fail with an AttributeError, and ignore
         __loader__.

       * In Python 3.12 and beyond, if you do not have a __loader__, we don't
         care as long as you still have a __spec__.loader.
    */
```

I'll work on an update.

**Comment by warsaw** [comment]
@brettcannon Please take another look.

File: Lib/importlib/_bootstrap_external.py [hunk 2]
```diff
@@ -862,6 +862,52 @@ def spec_from_file_location(name, location=None, *, loader=None,
     return spec
 
 
+def _bless_my_loader(module_globals):
+    """Helper function for _warnings.c
```
Comment (brettcannon) [inline Lib/importlib/_bootstrap_external.py:866]:
```suggestion
    """Helper function for _warnings.c .
```

File: Lib/importlib/_bootstrap_external.py [hunk 3]
```diff
@@ -862,6 +862,52 @@ def spec_from_file_location(name, location=None, *, loader=None,
...
+
+    missing = object()
+    loader = module_globals.get('__loader__', None)
+    spec = module_globals.get('__spec__', missing)
+
+    if loader is None and spec is missing:
+        #raise AttributeError('Module globals is missing a __spec__')
+        return None
+    if loader is None and spec is None:
+        raise ValueError('Module globals is missing a __spec__.loader')
```
Comment (brettcannon) [inline Lib/importlib/_bootstrap_external.py]:
```suggestion
    if loader is None:
        if spec is missing:
            # If working with a module:
            # raise AttributeError('Module globals is missing a __spec__')
            return None
        elif spec is None:
            raise ValueError('Module globals is missing a __spec__.loader')
```

File: Lib/importlib/_bootstrap_external.py [hunk 4]
```diff
@@ -862,6 +862,52 @@ def spec_from_file_location(name, location=None, *, loader=None,
...
+
+    spec_loader = getattr(spec, 'loader', missing)
+
+    if spec_loader in (missing, None):
```
Comment (brettcannon) [inline Lib/importlib/_bootstrap_external.py:894]:
```suggestion
    if spec_loader in {missing, None}:
```

File: Lib/importlib/_bootstrap_external.py [hunk 5]
```diff
@@ -862,6 +862,52 @@ def spec_from_file_location(name, location=None, *, loader=None,
...
+            raise exc('Module globals is missing a __spec__.loader')
+        _warnings.warn(
+            'Module globals is missing a __spec__.loader',
+            DeprecationWarning)
+        spec_loader = loader
+
+    assert spec_loader is not None
+    if loader is not None and loader != spec_loader:
+        _warnings.warn(
+            'Module globals; __loader__ != __spec__.loader',
```
Comment (brettcannon) [inline Lib/importlib/_bootstrap_external.py:906]:
```suggestion
            'Module globals' __loader__ != __spec__.loader',
```

File: Lib/test/test_importlib/import_/test_helpers.py [hunk 6]
```diff
@@ -67,5 +69,114 @@ def test_no_loader_no_spec_but_source(self):
...
+    # in favor of requiring __spec__.loader.
+
+    def test_gh86298_no_loader_and_no_spec(self):
+        bar = ModuleType('bar')
+        del bar.__loader__
+        del bar.__spec__
+        # 2022-10-06(warsaw): For backward compatibility with the
+        # implementation in _warnings.c, this can't raise an
+        # AttributeError.  See _bless_my_loader() in _bootstrap_external.py
+        ## self.assertRaises(
```
Comment (brettcannon) [inline Lib/test/test_importlib/import_/test_helpers.py:87]:
```suggestion
        # If working with a module:
        ## self.assertRaises(
```

File: Lib/test/test_importlib/import_/test_helpers.py [hunk 7]
```diff
@@ -67,5 +69,114 @@ def test_no_loader_no_spec_but_source(self):
...
+        self.assertIsNone(_bootstrap_external._bless_my_loader(bar.__dict__))
+
+    def test_gh86298_loader_is_none_and_no_spec(self):
+        bar = ModuleType('bar')
+        bar.__loader__ = None
+        del bar.__spec__
+        # 2022-10-06(warsaw): For backward compatibility with the
+        # implementation in _warnings.c, this can't raise an
+        # AttributeError.  See _bless_my_loader() in _bootstrap_external.py
+        ## self.assertRaises(
```
Comment (brettcannon) [inline Lib/test/test_importlib/import_/test_helpers.py:100]:
```suggestion
        # If working with a module:
        ## self.assertRaises(
```

File: Python/_warnings.c [hunk 8]
```diff
@@ -977,19 +977,26 @@ warnings_warn_impl(PyObject *module, PyObject *message, PyObject *category,
...
     PyObject *loader;
     PyObject *module_name;
     PyObject *get_source;
     PyObject *source;
     PyObject *source_list;
     PyObject *source_line;
 
-    /* Check/get the requisite pieces needed for the loader. */
-    loader = _PyDict_GetItemWithError(module_globals, &_Py_ID(__loader__));
+    /* stolen from import.c */
```
Comment (brettcannon) [inline Python/_warnings.c:988]:
```suggestion
    /* Stolen from import.c . */
```

**Review by brettcannon — APPROVED** [review]

**Pull request merged**